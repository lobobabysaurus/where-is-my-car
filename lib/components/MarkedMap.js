import { Location, MapView } from 'expo'
import PropTypes from 'prop-types'
import React from 'react'

import styles from '../styles'

export default class MarkedMap extends React.Component {

  state = {
    address: undefined,
  }

  componentDidMount = async () => {
    const {coords} = this.props
    this._updateAddress(coords)
  }

  componentDidUpdate = async (oldProps) => {
    const {coords} = this.props
    if (coords && (coords !== oldProps.coords)) {
      this._updateAddress(coords)
    }
  }

  render() {
    const {coords, region} = this.props
    const marker = coords
      ? <MapView.Marker
        coordinate={coords}
        title="Car"
        description={this._getMarkerDescription()} />
      : undefined
    return(
      <MapView style={styles.map} initialRegion={region}>
        {marker}
      </MapView>
    )
  }

  _getMarkerDescription = () => {
    let location = 'Here'
    if (this.state.address) {
      location = `On ${this.state.address.street}`
    }
    return `You Parked ${location}!`
  }

  _updateAddress = async (coords) => {
    if (coords) {
      const addresses = await Location.reverseGeocodeAsync(coords)
      if (addresses.length) {
        const address = addresses[0]
        this.setState({address})
      }
    }
  }
}

MarkedMap.propTypes = {
  coords: PropTypes.shape({
    latitude: PropTypes.number.isRequired,
    longitude: PropTypes.number.isRequired
  }),
  region: PropTypes.shape({
    latitude: PropTypes.number.isRequired,
    latitudeDelta: PropTypes.number.isRequired,
    longitude: PropTypes.number.isRequired,
    longitudeDelta: PropTypes.number.isRequired,
  }).isRequired,
}
