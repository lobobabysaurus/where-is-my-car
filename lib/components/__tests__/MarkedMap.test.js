import React from 'react'
import renderer from 'react-test-renderer'

import MarkedMap from '../MarkedMap'

jest.mock('react-native-maps', () => {
  const React = require.requireActual('react')
  const MapView = require.requireActual('react-native-maps')

  const MockCallout = (props) => (
    React.createElement('Callout', props, props.children)
  )

  const MockMarker = (props) => (
    React.createElement('Marker', props, props.children)
  )

  const MockMapView = (props) => (
    React.createElement('MapView', props, props.children)
  )

  MockCallout.propTypes = MapView.Callout.propTypes
  MockMarker.propTypes = MapView.Marker.propTypes
  MockMapView.propTypes = MapView.propTypes
  MockMapView.Marker = MockMarker
  MockMapView.Callout = MockCallout
  return MockMapView
})

describe('MarkedMap', () => {
  const region = {
    latitude: 0,
    latitudeDelta: 1,
    longitude: 0,
    longitudeDelta: 1,
  }

  test('renders correctly for the provided region with no markers by default', () => {
    const component = renderer.create(<MarkedMap region={region} />).toJSON()
    expect(component).toMatchSnapshot()
  })

  test('render a marker when a coordinate is provided', () => {
    const coords = {
      latitude: 5,
      longitude: 5,
    }
    const component = renderer.create(
      <MarkedMap region={region} coords={coords} />
    )
    expect(component).toMatchSnapshot()
  })
})
