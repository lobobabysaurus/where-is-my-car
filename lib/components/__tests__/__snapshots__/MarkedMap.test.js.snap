// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkedMap render a marker when a coordinate is provided 1`] = `
<MapView
  initialRegion={
    Object {
      "latitude": 0,
      "latitudeDelta": 1,
      "longitude": 0,
      "longitudeDelta": 1,
    }
  }
  style={
    Object {
      "flexBasis": "90%",
    }
  }
>
  <Marker
    coordinate={
      Object {
        "latitude": 5,
        "longitude": 5,
      }
    }
    description="You Parked Here!"
    title="Car"
  />
</MapView>
`;

exports[`MarkedMap renders correctly for the provided region with no markers by default 1`] = `
<MapView
  initialRegion={
    Object {
      "latitude": 0,
      "latitudeDelta": 1,
      "longitude": 0,
      "longitudeDelta": 1,
    }
  }
  style={
    Object {
      "flexBasis": "90%",
    }
  }
/>
`;
